Add ILogWriter interface to ilc and crossgen

vitek-karas opened the pull request:
This will allow test infra to capture full detail of reported diagnostics (not just text, but the actual type system objects).

Part of the work to bring linker tests over to Native AOT.

Note: the products don't use this in any way. The plan is for the tests to have a project reference to ILCompiler and use test-specific log writer through the APIs there.

vitek-karas committed:
Add ILogWriter interface to ilc and crossgen

This will allow test infra to capture full detail of reported diagnostics (not just text, but the actual type system objects).

Part of the work to bring linker tests over to Native AOT.

Note: the products don't use this in any way. The plan is for the tests to have a project reference to ILCompiler and use test-specific log writer through the APIs there.

tlakollo reviewed (approved):
Looks good, just as a comment why not hook it up with the Program.cs of each of the projects like in the original PR?

vitek-karas commented:
To me it makes sense to have a .ctor overload which knows how to write to text... by default. But I can change it.

tlakollo commented:
> To me it makes sense to have a .ctor overload which knows how to write to text... by default. But I can change it.
Gotcha, yes the .ctor overload looks like a better option

MichalStrehovsky commented on src/coreclr/tools/Common/Compiler/Logger.cs:
```diff
@@ -27,46 +29,56 @@ public class Logger
+        public static Logger Null = new Logger(new TextLogWriter(TextWriter.Null), false);
 
-        public TextWriter Writer { get; }
+        public TextWriter Writer => _logWriter.Writer;
```
On my phone so just asking here - would it be possible to migrate uses of this to LogMessage. Then maybe ILogWriter doesn't need the Writer property?

MichalStrehovsky reviewed (approved):
Thank you!

vitek-karas committed:
Rework the change - separate loggers for ilc and crossgen

In other changes where I'm tryhing to share more code with the linker and NativeAOT the idea of sharing the logger between crossgen and ilc becomes really problematic (the MessageOrigin and related classes become even more coupled with the ilc/linker code).
So I decided to split the two loggers completely, the ilc is really the only one using the MessageOrigin and related functionality, while crossgen uses the very simple string based methods only.

So I moved all of the ilc specific classes into the ilc directories along with the logger.
I then introduced a very simple logger class just for crossgen which implements only the necessary functionality for that tool.

If there's a will to share more in the future (no reason right now I'm aware of) we should design a better abstraction for the two tools to share.

vitek-karas commented:
I reworked the change to split loggers between ilc/crossgen. For the ilc version I kept the new abstraction `ILogWriter`, crossgen doesn't have it as it's not applicable.

In other changes where I'm trying to share more code with the linker and NativeAOT the idea of sharing the logger between crossgen and ilc becomes really problematic (the MessageOrigin and related classes become even more coupled with the ilc/linker code).
So I decided to split the two loggers completely, the ilc is really the only one using the MessageOrigin and related functionality, while crossgen uses the very simple string based methods only.

So I moved all of the ilc specific classes into the ilc directories along with the logger.
I then introduced a very simple logger class just for crossgen which implements only the necessary functionality for that tool.

If there's a will to share more in the future (no reason right now I'm aware of) we should design a better abstraction for the two tools to share.

MichalStrehovsky reviewed (approved):
Thank you!
